From ad068d827f1737c548683ed2764de32a575522b8 Mon Sep 17 00:00:00 2001
From: gongchangsui <gongchangsui@outlook.com>
Date: Mon, 19 Jan 2026 21:57:02 -0500
Subject: [PATCH] add total_fragment_block_num

---
 hw/char/virtio-console.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/hw/char/virtio-console.c b/hw/char/virtio-console.c
index ee19d50..f38bbe9 100644
--- a/hw/char/virtio-console.c
+++ b/hw/char/virtio-console.c
@@ -148,8 +148,8 @@ typedef struct {
 	int32_t ptzfd;
 	int32_t cpu_index;
 	int32_t err_flag;
-	int32_t is_fragment;
 	uint32_t fragment_block_num;
+	uint32_t total_fragment_block_num;
 	uint32_t vm_page_size;;
 	uint64_t block_addrs[TEEC_PARAM_NUM];//qemu and proxy don't use
 	uint32_t block_size[TEEC_PARAM_NUM];
@@ -163,8 +163,8 @@ typedef struct {
 	uint32_t seq_num;
 	int32_t ptzfd;
 	int32_t err_flag;
-	int32_t is_fragment;
 	uint32_t fragment_block_num;
+	uint32_t total_fragment_block_num;
 	uint32_t vm_page_size;;
 	uint64_t block_addrs[TEEC_PARAM_NUM];//qemu and proxy don't use
 	uint32_t block_size[TEEC_PARAM_NUM];
-- 
1.8.3.1

