From be59770ba018965f0024b1d9d1abf802791c2cff Mon Sep 17 00:00:00 2001
From: gongchangsui <gongchangsui@outlook.com>
Date: Sat, 29 Nov 2025 01:50:23 -0500
Subject: [PATCH] add vm uid

Signed-off-by: gongchangsui <gongchangsui@outlook.com>
---
 hw/char/tc_ns_client.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/hw/char/tc_ns_client.h b/hw/char/tc_ns_client.h
index 657a51b..47c846f 100644
--- a/hw/char/tc_ns_client.h
+++ b/hw/char/tc_ns_client.h
@@ -77,6 +77,7 @@ typedef struct {
             uint32_t file_h_addr;
         } memref;
     };
+    unsigned int vm_uid;
 } TC_NS_ClientContext;
 
 typedef struct {
-- 
1.8.3.1

